### PR TITLE
Fujifilm compressed RAF support.

### PR DESCRIPTION
**Pull request by muessel**
Hello,

this pull request implements support for Fujifilm compressed RAF Files for the X-Pro2, X-T2 and X-T20. The code is based on LibRaw / rawtherapee.

### CODE REVIEW

**Comment by LebedevRI** [comment]
Hi.
Nice to finally see that.
Are there *any* changes other than adapting to some of rs's infrastructure?

File: src/librawspeed/decompressors/FujiDecompressor.cpp [hunk 1]
```diff
@@ -0,0 +1,1097 @@
...
+    FujiDecompressor - Decompress Fujifilm compressed RAF.
+
+    Copyright (C) 2016 Alexey Danilchenko
+    Copyright (C) 2016 Alex Tutubalin
+    Copyright (C) 2017 Uwe Müssel
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
```
Comment (LebedevRI) [inline src/librawspeed/decompressors/FujiDecompressor.cpp]:
```diff
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
```
Hmm, where did you get that?
https://github.com/LibRaw/LibRaw/commit/aeab4d8ea58f85050d1253bdc41f51baa79ffe65#diff-03b018f1deeb8fc38d2c927f09465635R11 says:
```diff
+LibRaw is free software; you can redistribute it and/or modify
+it under the terms of the one of three licenses as you choose:
+
+1. GNU LESSER GENERAL PUBLIC LICENSE version 2.1
+   (See file LICENSE.LGPL provided in LibRaw distribution archive for details).
+
+2. COMMON DEVELOPMENT AND DISTRIBUTION LICENSE (CDDL) Version 1.0
+   (See file LICENSE.CDDL provided in LibRaw distribution archive for details).
+
```
(2 vs 2.1)

File: data/cameras.xml [hunk 2]
```diff
@@ -8598,6 +8598,19 @@
...
+		<ID make="Fujifilm" model="X-Pro2">Fujifilm X-Pro2</ID>
+		<CFA2 width="6" height="6">
+			<ColorRow y="0">GGRGGB</ColorRow>
+			<ColorRow y="1">GGBGGR</ColorRow>
+			<ColorRow y="2">BRGRBG</ColorRow>
+			<ColorRow y="3">GGBGGR</ColorRow>
+			<ColorRow y="4">GGRGGB</ColorRow>
+			<ColorRow y="5">RBGBRG</ColorRow>
+		</CFA2>
+		<Crop x="0" y="6" width="-16" height="0"/>
```
Comment (LebedevRI) [inline data/cameras.xml:8611]: Ugh, so they are even writing different raw image WxH in compressed vs. uncompressed formats..

File: src/librawspeed/decompressors/FujiDecompressor.cpp [hunk 3]
```diff
@@ -0,0 +1,1097 @@
...
+  }
+
+  *data <<= bits_left;
+  bits_left_in_byte -= bits_left;
+  *data |= ((1 << bits_left) - 1) &
+           ((unsigned)info->cur_buf[info->cur_pos] >> bits_left_in_byte);
+  info->cur_bit = (8 - (bits_left_in_byte & 7)) & 7;
+}
+
+int FujiDecompressor::bitDiff(int value1, int value2) {
```
Comment (LebedevRI) [inline src/librawspeed/decompressors/FujiDecompressor.cpp]:
```
int __attribute__(const) FujiDecompressor::bitDiff(int value1, int value2) {
```

File: src/librawspeed/decoders/RafDecoder.cpp [hunk 4]
```diff
@@ -155,6 +165,17 @@ RawImage RafDecoder::decodeRawInternal() {
...
   if (!this->checkCameraSupported(meta, mRootIFD->getID(), ""))
     ThrowRDE("Unknown camera. Will not guess.");
+
+  if (isCompressed()) {
+    mRaw->metadata.mode = "compressed";
+
+    auto id = mRootIFD->getID();
+    const Camera* cam = meta->getCamera(id.make, id.model, mRaw->metadata.mode);
+    if (!cam)
+      ThrowRDE("Couldn't find camera %s %s", id.make, id.model);
```
Comment (LebedevRI) [inline src/librawspeed/decoders/RafDecoder.cpp]:
```
ThrowRDE("Couldn't find camera %s %s", id.make.c_str(), id.model.c_str());
```

File: src/librawspeed/decompressors/FujiDecompressor.h [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+#include "LJpegDecompressor.h"
+
+namespace rawspeed {
+
+class FujiDecompressor //: public LJpegDecompressor
+{
+public:
+  FujiDecompressor(Buffer input, RawImage img, int offset);
+  ~FujiDecompressor(void);
+
```
Comment (LebedevRI) [inline src/librawspeed/decompressors/FujiDecompressor.h:34]: i guess you want to add
```
using ushort = ushort16;
```
right here for now.

**Review by LebedevRI — CHANGES_REQUESTED** [review]
Please make CI happy :)
The code is far from ideal, which is quite understandable, but it will be easier for me to fix it after merging as-is...

File: src/librawspeed/decompressors/FujiDecompressor.cpp (hunk 1 above)
Comment (muessel) [reply src/librawspeed/decompressors/FujiDecompressor.cpp]: Is it enough to change the version from 2 to 2.1 or do I have to copy the whole LibRaw license header?

File: data/cameras.xml (hunk 2 above)
Comment (muessel) [reply data/cameras.xml:8611]: Yes, uncomressed and comressed files are totally different, eg.:

exiftool Fujifilm\ -\ X-T20\ -\ compressed\ \(3-2\).RAF
...
Raw Image Width                 : 6032
Raw Image Height                : 4032
Raw Image Full Width            : 6048
Raw Image Full Height           : 4038

exiftool Fujifilm\ -\ X-T20\ -\ Uncompressed\ \(3-2\).RAF
...
Raw Image Width                 : 6032
Raw Image Height                : 4032
Raw Image Full Width            : 6160
Raw Image Full Height           : 4032

Comment (LebedevRI) [reply data/cameras.xml:8611]: Yes, that is what i was talking about.

File: src/librawspeed/decompressors/FujiDecompressor.cpp (hunk 1 above)
Comment (LebedevRI) [reply src/librawspeed/decompressors/FujiDecompressor.cpp]: I think it is enough just to change the version.

**Comment by LebedevRI** [comment]
Ok, i see.
Ignore those windows clang-4.0 warnings, i'll handle them manually.

**Comment by muessel** [comment]
Travis CI has problems with mac builds with std::abs. Do I have to change them to abs and then include "using std::abs;" ?

**Comment by LebedevRI** [comment]
Will fix locally, thank you!

File: data/cameras.xml (hunk 2 above)
Comment (LebedevRI) [reply data/cameras.xml:8611]: @muessel 
An entry for FUJIFILM X100F is missing.
Is there some problem with that camera, or it is a simple oversight?

**Comment by muessel** [comment]
Hello Roman,

I just forgot the the X100F also supports the compressed format.

I've seen you also implemented support for the GFX 50S. In the method
FujiDecompressor::copy_line_to_bayer
a static CFA is used.  Maybe this should be converted into something
configurable like with the X-Trans sensors.

Regards
   Uwe

2017-06-24 13:22 GMT+02:00 Roman Lebedev <notifications@github.com>:

**Comment by LebedevRI** [comment]
> Hello Roman,
> I just forgot the the X100F also supports the compressed format.

Yeah, i checked, and X100F compressed works too, fixed in branch.
Also, while there, FUJIFILM GFX 50S uncompressed+compressed support added - the only fuji camera with bayer sensor and with this compresson scheme, i believe.

What i'm currently missing is some compressed .RAF with 12-bits data, so this branch is taken:
https://github.com/darktable-org/rawspeed/blob/fae7f4014024f7aeb34fcab0c2d7967377764ca6/src/librawspeed/decompressors/FujiDecompressor.cpp#L106-L110
I think, it is produced in high-speed continuous shooting mode, 14/12/8/whaever fps.

**Comment by LebedevRI** [comment]
> I've seen you also implemented support for the GFX 50S. In the method
FujiDecompressor::copy_line_to_bayer
a static CFA is used.  Maybe this should be converted into something
configurable like with the X-Trans sensors.

All that is very strange. If that is the actual code used in current libraw, then it may mean two things (both can be valid at the same time):
1. All cameras which support compressed raw, have the exact same CFA filter pattern
2. The CFA used in compression is static/the same for all cameras/does not necessarily match the actual camera's CFA pattern.

**Comment by muessel** [comment]
>
> What i'm currently missing is some compressed .RAF with 12-bits data, so
> this branch is taken:
> https://github.com/darktable-org/rawspeed/blob/
> fae7f4014024f7aeb34fcab0c2d7967377764ca6/src/librawspeed/decompressors/
> FujiDecompressor.cpp#L106-L110
> I think, it is produced in high-speed continuous shooting mode,
> 14/12/8/whaever fps.
>
I've tested it on my X-T2 with 8fps mechanical shutter and 11/14 fps
electronical shutter. The .RAFs are all 14bit according to the code
mentioned above.

**Comment by LebedevRI** [comment]
https://discuss.pixls.us/t/compressed-raf-help-needed/4527

**Comment by Floessie** [comment]
@muessel 
> Travis CI has problems with mac builds with std::abs. Do I have to change them to abs and then include "using std::abs;" ?

Uwe, just my 2¢ (I know Roman will fix or has already fixed it): Most often the `abs()` problem isn't about the namespace, but a change from the way this was defined for C++03 vs. C++11. There's an integer version in [`<cstdlib>`](http://en.cppreference.com/w/cpp/numeric/math/abs) and a float version in [`<cmath>`](http://en.cppreference.com/w/cpp/numeric/math/fabs). As the latter link notes:

> Between C++11 and C++14, the standard erroneously required std::abs to have overloads for integer types returning double. This requirement was removed in C++17 by defect report 2735.

That causes trouble again and again, and that's most probably the case here also. The best solution usually is [to use `std::fabs()` for floats and to make sure, you don't feed `std::abs()` with an `unsigned`]( https://github.com/Beep6581/RawTherapee/issues/3339#issuecomment-268975964).

HTH,
Flössie

**Comment by LebedevRI** [comment]
Update: still no sign of that elusive 12bit sample, and no reply from @LibRaw.
I'll wait a few more days and proceed without it...

**Pull request merged**